nemo/collections/nlp/models/nlp_model.py

```diff
@@ -312,101 +241,238 @@ def setup(self, stage: str) -> None:
-                        num_replicas=app_state.data_parallel_size,
-                        rank=app_state.data_parallel_rank,
-                    )
-                    mp_dl = self._trainer.replace_sampler(self._train_dl, mp_sampler)
-                    self._train_dl = mp_dl
-                else:
-                    raise NotImplementedError(
-                        f'The BERT encoder: {self.bert_model} does not support model parallelism yet.'
-                    )
-
     def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        """ LightningModule hook that's used to save things in addition to model weights. """
+
         if hasattr(self, "bert_model") and isinstance(self.bert_model, MegatronBertEncoder):
             checkpoint['checkpoint_version'] = get_checkpoint_version()
         return None
 
-    @property
-    def input_module(self):
-        return self.bert_model
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        """ LightningModule hook that's used to restore things saved with on_save_checkpoint."""
 
-    @property
-    def output_module(self):
-        return self.classifier
+        if hasattr(self, "bert_model") and isinstance(self.bert_model, MegatronBertEncoder):
+            if get_checkpoint_version():
+                assert (
+                    checkpoint['checkpoint_version'] == get_checkpoint_version()
+                ), 'checkpoint version found on_load_checkpoint different than get_checkpoint_version'
+            else:
+                set_checkpoint_version(checkpoint['checkpoint_version'])
+                logging.info(f"Setting Megatron checkpoint version: {checkpoint['checkpoint_version']}")
+        return None
 
+    # no rank check as model parallel models need to be saved on data parallel rank 0
+    def save_to(self, save_path: str):
+        """
+        Saves model instance (weights and configuration) into .nemo file
+         You can use "restore_from" method to fully restore instance from .nemo file.
 
-class NLPCheckpointConnector(CheckpointConnector):
-    """ Override PTL CheckpointConnector to support model parallel checkpoints from Megatron-LM.
-    """
+        .nemo file is an archive (tar.gz) with the following:
+            model_config.yaml - model configuration in .yaml format. You can deserialize this into cfg argument for model's constructor
+            model_wights.chpt - model checkpoint
 
-    def __init__(self, trainer):
-        super().__init__(trainer)
+        Args:
+            save_path: Path to .nemo file where model instance should be saved
+        """
 
-    def save_checkpoint(self, filepath, weights_only: bool):
-        """Slightly modified version of PyTorch Lightning's save_checkpoint.
+        app_state = AppState()
+        if app_state.model_parallel_size is not None:
+            self._default_save_to(save_path)
+        else:
+            return super().save_to(save_path)
```

Comment:
Could you add a comment here saying this super()... line will execute only on rank 0?